src/main/java/com/minecolonies/coremod/colony/buildings/modules/AbstractCraftingBuildingModule.java

```diff
@@ -67,8 +220,481 @@ public void serializeToView(PacketBuffer buf)
+                {
+                    addRecipeToList(recipeToken, true);
+                    building.getColony().getRequestManager().onColonyUpdate(request -> request.getRequest() instanceof IDeliverable && ((IDeliverable) request.getRequest()).matches(recipeStorage.getPrimaryOutput()));
+                    markDirty();
+                }
+                else if((forceReplace || newRecipe.getMustExist()) && !(duplicateFound.equals(recipeToken)))
+                {
+                    //We found the base recipe for a multi-recipe, replace it with the multi-recipe
+                    replaceRecipe(duplicateFound, recipeToken);
+                    building.getColony().getRequestManager().onColonyUpdate(request -> request.getRequest() instanceof IDeliverable && ((IDeliverable) request.getRequest()).matches(recipeStorage.getPrimaryOutput()));
+
+                    //Clean up old 'classic' recipes that the new multi-recipe replaces
+                    final List<ItemStack> alternates = recipeStorage.getAlternateOutputs();
+                    for(IToken<?> token : recipes)
+                    {
+                        final IRecipeStorage storage = recipeManager.getRecipes().get(token);
+                        if(storage.getRecipeType() instanceof ClassicRecipe && ItemStackUtils.compareItemStackListIgnoreStackSize(alternates, storage.getPrimaryOutput(), false, true))
+                        {
+                            removeRecipe(token);
+                        }
+                    }
+                    building.getColony().getRequestManager().onColonyUpdate(request -> request.getRequest() instanceof IDeliverable && recipeStorage.getAlternateOutputs().stream().anyMatch(i -> ((IDeliverable) request.getRequest()).matches(i)));
+                    markDirty();
+                }
+            }
+            else
+            {
+                if(recipes.contains(recipeToken))
+                {
+                    removeRecipe(recipeToken);
+                    markDirty();
+                }
+            }
+        }
+    }
+
+    @Override
+    public void clearRecipes()
+    {
+        recipes.clear();
+    }
+
+    /**
+     * @param token the token of the recipe storage.
+     * @return whether the recipe can bee added based on tokens.
+     */
+    protected Optional<Boolean> canRecipeBeAddedBasedOnTags(final IToken<?> token)
+    {
+        final IRecipeStorage storage = IColonyManager.getInstance().getRecipeManager().getRecipes().get(token);
+        return canRecipeBeAddedBasedOnTags(storage);
     }
 
+    /**
+     * @param storage the recipe storage to check.
+     * @return whether the recipe can bee added based on tokens.
+     */
+    protected Optional<Boolean> canRecipeBeAddedBasedOnTags(final IRecipeStorage storage)
+    {
+        if (storage == null)
+        {
+            return Optional.of(false);
+        }
+
+        final String crafterName = building.getJobName().toLowerCase();

```

Comment:
Given the tag definitions, presumably this needs to sometimes have a `_smelting` suffix (e.g. glassblower), and so probably needs to be calling a virtual method only overridden in those specific cases?

Reply:
This is quite funny. I think we need input of @Mekle001 here. Most buildings use CraftingUtils.isRecipeCompatibleBasedOnTags which I'm already calling the right way for the right workers.

This method is only called by a) Cook and b) Farmer, by the cook in the getFirstRecipe and by the farmer in the canAddRecipe code. Do we need this code at all?

Reply:
Actually all the modules call this, which means that they're checking the tags twice, which is probably wrong.  This is likely an incorrect translation when moving this from the building to the module.

Reply:
Sorry, no, I was searching the wrong method.  You were right, it's just the farmer and cook.  The farmer looks like a merge derp (reviewed separately).  It looks like the usage of the cook should be replaced with `isRecipeCompatibleWithCraftingModule`.  And then this code can go away.

Reply:
Talked with Mekle, yes, we can remove this.